src/Buckle/Buckle/CodeAnalysis/Evaluator.cs

```diff
@@ -67,10 +119,42 @@ internal sealed class Evaluator {
             return null;
 
         var body = LookupMethod(function);
-        return EvaluateStatement(body);
+        var result = EvaluateStatement(body);
+
+        if (result.isReference)
+            return GetVariableValue(result.reference);
+        else
+            return result.value;
+    }
+
+    private object GetVariableValue(VariableSymbol variable) {
+        EvaluatorObject value = null;
+
+        if (variable.type == SymbolType.GlobalVariable) {
+            value = globals_[variable];
+        } else {
+            var locals = locals_.Peek();
+            value = locals[variable];
+        }
+
+        if (value.isReference)
+            return GetVariableValue(value.reference);
+        else if (value.value is EvaluatorObject)
+            return Value(value.value as EvaluatorObject);
+        else
+            return value.value;
     }
 
-    private object EvaluateStatement(BoundBlockStatement statement) {
+    private object Value(EvaluatorObject value) {
+        if (value.isReference)
+            return GetVariableValue(value.reference);
+        else if (value.value is EvaluatorObject)
+            return Value(value.value as EvaluatorObject);
+        else
+            return value.value;
+    }
```

Comment:
Value and GetVariableValue are new helper functions that recursively get the actual value of an EvaluatorObject or variable. Just returns `this.value` unless either is a reference. In that case it looks at what it is pointing to and repeats this process.